render/vulkan/texture.c

```diff
@@ -0,0 +1,715 @@
+	img_info.samples = VK_SAMPLE_COUNT_1_BIT;
+	img_info.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+	img_info.initialLayout = VK_IMAGE_LAYOUT_PREINITIALIZED;
+	img_info.extent = (VkExtent3D) { width, height, 1 };
+	img_info.usage = VK_IMAGE_USAGE_SAMPLED_BIT;
+
+	VkImageLayout layout = VK_IMAGE_LAYOUT_GENERAL;
+	img_info.tiling = VK_IMAGE_TILING_OPTIMAL;
+	img_info.usage |= VK_IMAGE_USAGE_TRANSFER_DST_BIT;
+	mem_bits = vulkan_find_mem_type(renderer->dev,
+		VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT, mem_bits);
+	layout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+
+	res = vkCreateImage(dev, &img_info, NULL, &texture->image);
+	if (res != VK_SUCCESS) {
+		wlr_vk_error("vkCreateImage failed", res);
+		goto error;
+	}
+
+	// memory
+	VkMemoryRequirements mem_reqs;
+	vkGetImageMemoryRequirements(dev, texture->image, &mem_reqs);
+
+	VkMemoryAllocateInfo mem_info = {0};
+	mem_info.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+	mem_info.allocationSize = mem_reqs.size;
+	mem_info.memoryTypeIndex = mem_bits & mem_reqs.memoryTypeBits;
+	res = vkAllocateMemory(dev, &mem_info, NULL, &texture->memories[0]);
+	if (res != VK_SUCCESS) {
+		wlr_vk_error("vkAllocatorMemory failed", res);
+		goto error;
+	}
+
+	texture->mem_count = 1;
+	res = vkBindImageMemory(dev, texture->image, texture->memories[0], 0);
+	if (res != VK_SUCCESS) {
+		wlr_vk_error("vkBindMemory failed", res);
+		goto error;
+	}
+
+	// view
+	VkImageViewCreateInfo view_info = {0};
+	view_info.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
+	view_info.viewType = VK_IMAGE_VIEW_TYPE_2D;
+	view_info.format = texture->format->vk_format;
+	view_info.components.r = VK_COMPONENT_SWIZZLE_IDENTITY;
+	view_info.components.g = VK_COMPONENT_SWIZZLE_IDENTITY;
+	view_info.components.b = VK_COMPONENT_SWIZZLE_IDENTITY;
+	view_info.components.a = VK_COMPONENT_SWIZZLE_IDENTITY;
+
+	view_info.subresourceRange = (VkImageSubresourceRange) {
+		VK_IMAGE_ASPECT_COLOR_BIT, 0, 1, 0, 1
+	};
+	view_info.image = texture->image;
+
+	res = vkCreateImageView(dev, &view_info, NULL,
+		&texture->image_view);
+	if (res != VK_SUCCESS) {
+		wlr_vk_error("vkCreateImageView failed", res);
+		goto error;
+	}
+
+	// descriptor
+	texture->ds_pool = vulkan_alloc_texture_ds(renderer, &texture->ds);
+	if (!texture->ds_pool) {
+		wlr_log(WLR_ERROR, "failed to allocate descriptor");
+		goto error;
+	}
+
+	VkDescriptorImageInfo ds_img_info = {0};
+	ds_img_info.imageView = texture->image_view;
+	ds_img_info.imageLayout = layout;
+
+	VkWriteDescriptorSet ds_write = {0};
+	ds_write.sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
+	ds_write.descriptorCount = 1;
+	ds_write.descriptorType = VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER;
+	ds_write.dstSet = texture->ds;
+	ds_write.pImageInfo = &ds_img_info;
+
+	vkUpdateDescriptorSets(dev, 1, &ds_write, 0, NULL);
+
+	VkCommandBuffer cb = vulkan_record_stage_cb(renderer);
+	vulkan_change_layout(cb, texture->image,
+		VK_IMAGE_LAYOUT_PREINITIALIZED, VK_PIPELINE_STAGE_HOST_BIT,
+		VK_ACCESS_HOST_WRITE_BIT,
```

Comment:
Change this to VK_ACCESS_MEMORY_WRITE_BIT, with a layout of VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL once you've changed the way you upload textures.